File: android/src/main/kotlin/com/algorigo/in_app_updater/InAppUpdateInstallState.kt
package com.algorigo.in_app_updater

import com.google.android.play.core.install.InstallState
import com.google.android.play.core.install.model.InstallStatus
import com.google.android.play.core.ktx.bytesDownloaded
import com.google.android.play.core.ktx.hasTerminalStatus
import com.google.android.play.core.ktx.installStatus
import com.google.android.play.core.ktx.packageName
import com.google.android.play.core.ktx.totalBytesToDownload

class InAppUpdateInstallState(
  val installState: InstallState
) {

  fun toMap() = mapOf(
    "packageName" to installState.packageName,
    "installStatus" to installState.installStatus,
    "isDownloaded" to isDownloaded(),
    "bytesDownloaded" to installState.bytesDownloaded,
    "totalBytesToDownload" to installState.totalBytesToDownload,
    "hasTerminalStatus" to installState.hasTerminalStatus,
    "installErrorCode" to installState.installErrorCode()
  )

  fun isDownloaded() = installState.installStatus() == InstallStatus.DOWNLOADED
}

File: lib/data/in_app_update_install_state.dart
import 'package:in_app_updater/data/install_status.dart';
import 'package:in_app_updater/extensions/in_app_update_extensions.dart';

class InAppUpdateInstallState {
  final String? packageName;
  final InstallStatus? installStatus;
  final bool? isDownloaded;
  final int? bytesDownloaded;
  final int? totalBytesToDownload;
  final bool? hasTerminalStatus;
  final int? installErrorCode;

  InAppUpdateInstallState({
    required this.packageName,
    required this.installStatus,
    required this.isDownloaded,
    required this.bytesDownloaded,
    required this.totalBytesToDownload,
    required this.hasTerminalStatus,
    required this.installErrorCode,
  });

  factory InAppUpdateInstallState.fromJson(Map<String, dynamic> json) {
    return InAppUpdateInstallState(
      packageName: json['packageName'],
      installStatus: (json['installStatus'] as int?)?.toInstallStatus(),
      isDownloaded: json['isDownloaded'],
      bytesDownloaded: json['bytesDownloaded'],
      totalBytesToDownload: json['totalBytesToDownload'],
      hasTerminalStatus: json['hasTerminalStatus'],
      installErrorCode: json['installErrorCode'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'packageName': packageName,
      'installStatus': installStatus,
      'isDownloaded': isDownloaded,
      'bytesDownloaded': bytesDownloaded,
      'totalBytesToDownload': totalBytesToDownload,
      'hasTerminalStatus': hasTerminalStatus,
      'installErrorCode': installErrorCode,
    };
  }
}
